shell/platform/android/io/flutter/plugin/common/JSONCodec.java

```diff
@@ -0,0 +1,79 @@
+package io.flutter.plugin.common;
+
+import java.nio.ByteBuffer;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+/**
+ * A {@link MessageCodec} using UTF-8 encoded JSON messages, and supporting the same Java values
+ * as {@link JSONObject#wrap(Object)}.
+ */
+final class JSONCodec implements MessageCodec {
+    @Override
+    public ByteBuffer encodeMessage(Object message) {
+        if (message == null) {
+            return null;
+        }
+        return ByteBufferTools.extractBytes(JSONObject.wrap(message).toString());
+    }
+
+    @Override
+    public Object decodeMessage(ByteBuffer message) {
+        if (message == null) {
+            return null;
+        }
+        try {
+            final String json = ByteBufferTools.extractString(message);
+            final JSONTokener tokener = new JSONTokener(json);
+            final Object value = tokener.nextValue();
+            if (tokener.more()) {
+                throw new IllegalArgumentException("Invalid JSON");
+            }
+            return value;
+        } catch (JSONException e) {
+            throw new IllegalArgumentException("Invalid JSON", e);
+        }
+    }
+
+    @Override
+    public MethodCall decodeMethodCall(ByteBuffer message) {
```

Comment:
Separating the plain message codec from how to encode and decode method calls and result envelopes would probably imply a two-step decoding: first decode a message, then interpret that message as a method call or an envelope (and symmetrically for encoding). But that means that the method call and envelope must be messages too.

For JSON it doesn't matter that much, I guess, since the overhead is big anyway, but for the standard binary encoding, it might force us to represent a method call as a two-element list [method name, args] and a result envelope as a one element list for success and three-element list for error, or similar. The list objects must be constructed at both ends. But by coupling these things closer together, we can do method calls as two consecutive values with no list involved, and the result envelope using a single byte to distinguish the success and error cases.